relations/src/shielder/circuit_utils.rs

```diff
@@ -0,0 +1,69 @@
+use core::borrow::Borrow;
+#[cfg(feature = "std")]
+use std::fmt::{Display, Formatter};
+
+use ark_r1cs_std::{
+    alloc::{AllocVar, AllocationMode},
+    boolean::Boolean,
+    R1CSVar,
+};
+use ark_relations::r1cs::{Namespace, SynthesisError};
+
+use crate::CircuitField;
+
+#[derive(Clone, Debug)]
+pub(super) struct PathShapeVar {
+    shape: Vec<Boolean<CircuitField>>,
+}
+
+#[cfg(feature = "std")]
+impl Display for PathShapeVar {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{:?}",
+            self.shape
+                .iter()
+                .map(|b| b.value().map(|boo| if boo { "left" } else { "right" }))
+                .collect::<Vec<_>>()
+        )
+    }
+}
+
+impl PathShapeVar {
+    pub(super) fn len(&self) -> usize {
+        self.shape.len()
+    }
+
+    pub(super) fn at(&self, i: usize) -> &Boolean<CircuitField> {
```

Comment:
nit: one could implement [`Index`](https://doc.rust-lang.org/std/ops/trait.Index.html) and get the `[m]` access to `PathShapeVar` elements.

Reply:
done